### PR TITLE
Pokemon Optimizer round 4

### PR DESCRIPTION
**Pull request by Anakin5**
- Add full documentation of Pokemon Optimizer (need to add a few things but mostly done)
- Add support for different mode of selection "by_pokemon", "by_family", "overall"
- Add parameter to create and reference groups of Pokemon
- Add possibility to exclude Pokemon or groups of Pokemon
- Add possibility to configure evolution threshold when selecting Pokemon to keep
- Add possibility to keep Pokemon whose stats are no further that x% of the best
- Auto correct misspelled Pokemon names in configuration

### CODE REVIEW

**Comment by mjmadsen** [comment]
:+1:

**Comment by Cptspal** [comment]
I wonder if lines 261 to 263 of pokemon_optimizer.py are not a duplicate of lines above ? I've a TypeError 'Pokemon' type not iterable when I reach this step.

**Comment by Anakin5** [comment]
Which lines are you referring to ?

**Comment by mjmadsen** [comment]
I think he's referring to this:
            for _, pokemon_list in self.group_by_pokemon_id(senior_pokemon_list):
                try_evolve, keep = self.get_best_pokemon_for_rule(pokemon_list, rule)
                try_evolve_all += try_evolve
                keep_all += keep

            try_evolve, keep = self.sort_pokemon_list(other_family_list, rule)
            try_evolve_all += try_evolve
            keep_all += keep

File: pokemongo_bot/cell_workers/pokemon_optimizer.py [hunk 1]
```diff
@@ -62,211 +60,318 @@ def work(self):
...
+                keep_all += keep
+        elif len(senior_pids) < nb_branch:
+            # We did not get every combination yet = All other Pokemon are potentially good to keep
+            for _, pokemon_list in self.group_by_pokemon_id(senior_pokemon_list):
+                try_evolve, keep = self.get_best_pokemon_for_rule(pokemon_list, rule)
+                try_evolve_all += try_evolve
+                keep_all += keep
+
+            try_evolve, keep = self.sort_pokemon_list(other_family_list, rule)
+            try_evolve_all += try_evolve
```
Comment (mjmadsen) [inline pokemongo_bot/cell_workers/pokemon_optimizer.py]: Should have just commented here instead.

**Comment by Anakin5** [comment]
Sorry, can you point me on which line is reported in the error message ? I do not see where the error is.

**Comment by Cptspal** [comment]
I don't have the error on my screen anymore, but it was pointing to line 262 of pokemon_optimizer.py.
I've just set the evolve=false on config.json to get rid of the error, but now I transfert most of my pokemon so even if I set it back to true to check the error message, I've to wait for more pokemon.

Bit if you check the file pokemon_optimizer.py, you can see that line 261, 262 and 263 are just a duplicating of above lines 257, 258, 259 but outside of the "for" loop. I don't catch all the logic behind the code, so I may be wrong.

**Comment by Anakin5** [comment]
Ok I see the error, I will fix.

For the lines you are talking about, they are not exactly duplicated, there is a small difference ;)

**Comment by Anakin5** [comment]
I pushed the fix in the other pull request.

**Comment by mjmadsen** [comment]
I'm not seeing the PR

**Pull request merged**